### ISSUE TITLE
Rules do not work with named exports

### ISSUE THREAD

**Issue by Zamiell** [eslint-community/eslint-plugin-eslint-plugin#375]
Hello, and thanks for the plugin.

I tried to use these rules on my plugin today, but none of them seem to work, meaning that none of the rules were triggered.
I think this is because this plugin only looks for code matching `export default foo` or `module.exports = foo`.

My plugin does not use these and instead uses named exports. My plugin is a modern ESLint plugin written in TypeScript, and [using default exports in TypeScript is considered an anti-pattern](https://basarat.gitbook.io/typescript/main-1/defaultisbad).

For example, here is what [one of my rules](https://github.com/IsaacScript/isaacscript/blob/main/packages/eslint-plugin-isaacscript/src/rules/no-explicit-array-loops.ts) looks like:

```ts
import { ESLintUtils } from "@typescript-eslint/utils";

const createRule = ESLintUtils.RuleCreator(
  (ruleName) =>
    `https://github.com/IsaacScript/isaacscript/blob/main/packages/eslint-plugin-isaacscript/docs/rules/${ruleName}.md`,
);

export type Options = [];
export type MessageIds = "noExplicitArray";

export const noExplicitArrayLoops = createRule<Options, MessageIds>({
  name: "no-explicit-array-loops",
  meta: {
    type: "problem",
    docs: {
      description: "Disallows explicit iteration for arrays",
      recommended: "recommended",
      requiresTypeChecking: true,
    },
    schema: [],
    messages: {
      noExplicitArray:
        'Explicit iteration over arrays is disallowed. (Get rid of the "Object.values()".)',
    },
    fixable: "code",
  },
  defaultOptions: [],
  create(context) {
    // Rule logic here.
  },
});
```

Thus, I propose that the "detect rule" logic of this plugin is updated in some way. If you can advise on what the correct solution is, I can submit a pull request, as I would love to use the rules that this project offers. Thank you.

**Comment by bmish** [comment]
Thanks for your interest in using and contributing to our plugin!

Here's why we don't support your setup yet:

1. While we support the `ESLintUtils.RuleCreator` helper when used directly, we don't yet support it when the function is stored in a variable. I would like to support this, but it would require additional static analysis (maybe our `findVariableValue` helper) to detect the variable and follow its usage, which can get a bit complicated, and we generally only bother with it for common use cases. PR welcome.
2. We currently only look for default-exported rules because:
    1. At least in the traditional ESLint / CJS, default exports made it easier to just do `'foo-rule-name': require('foo-rule-name')` when exporting the rule objects for ESLint, without having to deal with transforming between rule names and variable names. So my impression is that default exports were pretty universal. In ESM, this benefit isn't as relevant, so perhaps named exports are fair-game / more common now. I see you have a [script](https://github.com/IsaacScript/isaacscript/blob/9c7c307faaa5c78efa1e8b96becc86d6816a86c0/packages/eslint-plugin-isaacscript/src/rules.ts) to generate the import/export index of rules that accommodates named exports, which is cool.
    2. Given the above assumption that almost everyone in the ESLint CJS world used default exports, ignoring named exports helped reduce false positives of detecting something that looked like a rule but wasn't.
    3. So overall, it sounds like supporting named exports is a change we could make. We can always use heuristics like limiting this support to ESM/TypeScript or even to the presence of certain known rule helpers to minimize the chance of false positives.

Here you can see the logic for how we detect rules and test case for the various setups we support: 

https://github.com/eslint-community/eslint-plugin-eslint-plugin/blob/c308547450f20a6aa4d80c058fa8568dd10a1f6c/lib/utils.js#L320

https://github.com/eslint-community/eslint-plugin-eslint-plugin/blob/c308547450f20a6aa4d80c058fa8568dd10a1f6c/tests/lib/utils.js#L13

**Comment by Zamiell** [comment]
@bmish Thanks for the reply. I took a look at I think this task is a little over my head. Can I request that someone from your core team implements this feature? I think it will be important for the ecosystem going forward, especially as more ESLint plugins switch to ESM & the new config format.

**Comment by aladdin-add** [comment]
@Zamiell no worries, I'll take care of it.

**Comment by 43081j** [comment]
@aladdin-add did you ever get around to this?

i use named exports in the various plugins i maintain, and have done for a long time now. so this plugin isn't of much use to me

if you don't have time, i can do the PR i'm sure. just let me know (although i'll probably go do it now anyway unless you already have one open)
